jax/_src/dtypes.py

```diff
@@ -343,7 +343,10 @@ def dtype(x, *, canonicalize=False):
   elif type(x) in python_scalar_dtypes:
     dt = python_scalar_dtypes[type(x)]
   else:
-    dt = np.result_type(x)
+    try:
+      dt = np.result_type(x)
+    except TypeError:
+      dt = np.asarray(x).dtype
```

Comment:
Is this change used by this PR? What's the case this line is meant to handle?

Reply:
Yes - we now pass memoryview objects to `_lattice_result_type`, and causes a TypeError in `np.result_type`.

Reply:
I ended up reverting this change and handling it in the `array` function. If `np.result_type` raises a TypeError for an input, then `jnp.result_type` should probably do so as well.